From 7c38bbaa7e84eb8c7c48155bd37293b210cd82e1 Mon Sep 17 00:00:00 2001
From: Yong Zhi <yong.zhi@intel.com>
Date: Sun, 6 May 2018 22:40:54 -0500
Subject: [PATCH 032/129] media: intel-ipu4: psys: Convert init_timer() to use
 timer_setup()

Message for Open Source:
Wrap code with kernel version checking to support updated timer APIs.

Message for Internal:
[Issue/Feature] The kernel timer interface has number of issues detailed
in this article <https://lwn.net/Articles/735887>.
[Root Cause/Changes]
Follow media/v4.15-1 treewide conversion.

Tracked-On: #JKCCIB-201
Change-Id: I3a7c8c1dd8683ad06273667a0ec8f3cbe89cde71
Signed-off-by: Yong Zhi <yong.zhi@intel.com>
Signed-off-by: Meng Wei <wei.meng@intel.com>
---
 drivers/media/pci/intel/ipu4/ipu4-psys.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/drivers/media/pci/intel/ipu4/ipu4-psys.c b/drivers/media/pci/intel/ipu4/ipu4-psys.c
index 82fa987..d183008 100644
--- a/drivers/media/pci/intel/ipu4/ipu4-psys.c
+++ b/drivers/media/pci/intel/ipu4/ipu4-psys.c
@@ -544,9 +544,15 @@ static int ipu_psys_kcmd_start(struct ipu_psys *psys,
 	return ret;
 }
 
+#if LINUX_VERSION_CODE <= KERNEL_VERSION(4, 14, 2)
 static void ipu_psys_watchdog(unsigned long data)
 {
 	struct ipu_psys_kcmd *kcmd = (struct ipu_psys_kcmd *)data;
+#else
+static void ipu_psys_watchdog(struct timer_list *t)
+{
+	struct ipu_psys_kcmd *kcmd = from_timer(kcmd, t, watchdog);
+#endif
 	struct ipu_psys *psys = kcmd->fh->psys;
 
 	queue_work(IPU_PSYS_WORK_QUEUE, &psys->watchdog_work);
@@ -693,9 +699,13 @@ int ipu_psys_kcmd_new(struct ipu_psys_command *cmd, struct ipu_psys_fh *fh)
 
 	ipu_psys_resource_alloc_init(&kcmd->resource_alloc);
 
+#if LINUX_VERSION_CODE <= KERNEL_VERSION(4, 14, 2)
 	init_timer(&kcmd->watchdog);
 	kcmd->watchdog.data = (unsigned long)kcmd;
 	kcmd->watchdog.function = &ipu_psys_watchdog;
+#else
+	timer_setup(&kcmd->watchdog, ipu_psys_watchdog, 0);
+#endif
 
 	if (cmd->min_psys_freq) {
 		kcmd->constraint.min_freq = cmd->min_psys_freq;
-- 
1.9.1

